piper/mousemap.py

```diff
@@ -0,0 +1,390 @@
+
+    def __init__(self, widget, is_left, svg_id):
+        self._widget = widget
+        self._is_left = is_left
+        self._svg_id = svg_id
+        self._svg_leader = svg_id + "-leader"
+
+    @property
+    def widget(self):
+        # The widget belonging to this child.
+        return self._widget
+
+    @property
+    def svg_id(self):
+        # The identifier of the SVG element with which this child's widget is
+        # paired.
+        return self._svg_id
+
+    @property
+    def svg_leader(self):
+        # The identifier of the leader SVG element with which this child's
+        # widget is paired.
+        return self._svg_leader
+
+    @property
+    def is_left(self):
+        # True iff this child's widget is allocated to the left of the SVG.
+        return self._is_left
+
+
+class MouseMap(Gtk.Container):
+    """A Gtk.Container subclass to draw a device SVG with child widgets that
+    map to the SVG. The SVG should have objects with identifiers, whose value
+    should also be given to the `do_add` method. See `do_size_allocate` and
+    https://github.com/libratbag/libratbag/blob/master/data/README.md for more
+    information.
+    """
+
+    __gtype_name__ = "MouseMap"
+
+    __gproperties__ = {
+        "spacing": (int,
+                    "spacing",
+                    "The amount of space between children and the SVG leaders",
+                    0, GLib.MAXINT, 0,
+                    GObject.PARAM_READABLE),
+    }
+
+    def __init__(self, layer, ratbagd_device, spacing=10, *args, **kwargs):
+        """Instantiates a new MouseMap.
+
+        @param layer The SVG layer whose leaders to draw, according to librsvg
+                     so e.g. `#layer1`.
+        @param ratbagd_device The device that should be mapped, as
+                              ratbagd.RatbagdDevice
+        @param spacing The spacing to place between the SVG's leaders and the
+                       widgets, as int
+
+        @raises ValueError when an argument is invalid, or when the given device
+                           has no image registered.
+        @raises GLib.Error when the SVG cannot be loaded.
+        """
+        if layer is None:
+            raise ValueError("Layer cannot be None")
+        if ratbagd_device is None:
+            raise ValueError("Device cannot be None")
+        if not os.path.isfile(ratbagd_device.svg_path):
+            raise ValueError("Device has no image or its path is invalid")
+
+        Gtk.Container.__init__(self, *args, **kwargs)
+        self.set_has_window(False)
+
+        self.spacing = spacing
+        self._layer = layer
+        self._device = ratbagd_device
+        self._children = []
+        self._highlight_element = None
+
+        self._handle = Rsvg.Handle.new_from_file(ratbagd_device.svg_path)
+        self._svg_data = etree.parse(ratbagd_device.svg_path)
+
+    def do_add(self, widget):
+        """Not implemented, use `add(widget, svg_id)` instead."""
+        pass
```

Comment:
In fact, Python will automatically reference the `add(widget, svg_id)` method: 

```
 > ./piper.in
Running from source tree, using local files
Traceback (most recent call last):
  File "/home/jente/code/src/github.com/libratbag/piper/piper/application.py", line 53,
in do_activate
    window = Window(self, self._ratbag)
  File "/home/jente/code/src/github.com/libratbag/piper/piper/window.py", line 102, in _
_init__
    stack.add_titled(self._setup_buttons_page(device), "buttons", _("Buttons"))
  File "/home/jente/code/src/github.com/libratbag/piper/piper/window.py", line 145, in _
setup_buttons_page
    mousemap.add(mapbutton)
TypeError: add() missing 1 required positional argument: 'svg_id'
```

To use the virtual `do_add` method and fall through to the line in my earlier comment, the user (developer) will have to explicitly call `mousemap.do_add(mousemap, widget)` which is unlikely to happen.

So I'll just remove `do_add` and we should be fine.